src/java/org/apache/cassandra/index/sai/StorageAttachedIndexBuilder.java

```diff
@@ -0,0 +1,351 @@
+import org.apache.cassandra.io.util.RandomAccessReader;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.Throwables;
+import org.apache.cassandra.utils.TimeUUID;
+import org.apache.cassandra.utils.concurrent.CountDownLatch;
+import org.apache.cassandra.utils.concurrent.Ref;
+
+import static org.apache.cassandra.utils.TimeUUID.Generator.nextTimeUUID;
+
+/**
+ * Multiple storage-attached indexes can start building concurrently. We need to make sure:
+ * 1. Per-SSTable index files are built only once
+ *      a. Per-SSTable index files already built, do nothing
+ *      b. Per-SSTable index files are currently building, we need to wait until it's built in order to consider index built.
+ * 2. Per-column index files are built for each column index
+ */
+public class StorageAttachedIndexBuilder extends SecondaryIndexBuilder
+{
+    protected static final Logger logger = LoggerFactory.getLogger(StorageAttachedIndexBuilder.class);
+
+    // make sure only one builder can write to per sstable files when multiple storage-attached indexes are created simultaneously.
+    private static final Map<SSTableReader, CountDownLatch> inProgress = Maps.newConcurrentMap();
+
+    private final StorageAttachedIndexGroup group;
+    private final TableMetadata metadata;
+    private final Tracker tracker;
+    private final TimeUUID compactionId = nextTimeUUID();
+    private final boolean isFullRebuild;
+    private final boolean isInitialBuild;
+
+    private final SortedMap<SSTableReader, Set<StorageAttachedIndex>> sstables;
+
+    private long bytesProcessed = 0;
+    private final long totalSizeInBytes;
+
+    StorageAttachedIndexBuilder(StorageAttachedIndexGroup group,
+                                SortedMap<SSTableReader, Set<StorageAttachedIndex>> sstables,
+                                boolean isFullRebuild,
+                                boolean isInitialBuild)
+    {
+        this.group = group;
+        this.metadata = group.metadata();
+        this.sstables = sstables;
+        this.tracker = group.table().getTracker();
+        this.isFullRebuild = isFullRebuild;
+        this.isInitialBuild = isInitialBuild;
+        this.totalSizeInBytes = sstables.keySet().stream().mapToLong(SSTableReader::uncompressedLength).sum();
+    }
+
+    @Override
+    public void build()
+    {
+        logger.debug(logMessage("Starting full index build"));
+        for (Map.Entry<SSTableReader, Set<StorageAttachedIndex>> e : sstables.entrySet())
+        {
+            SSTableReader sstable = e.getKey();
+            Set<StorageAttachedIndex> indexes = e.getValue();
+
+            Set<StorageAttachedIndex> existing = validateIndexes(indexes, sstable.descriptor);
+            if (existing.isEmpty())
+            {
+                logger.debug(logMessage("{} dropped during index build"), indexes);
+                continue;
```

Comment:
Trying to think through if we should `continue` or `break` here. If there are no `existing` indexes on one SSTable in this table, are indexes going to just appear for the other SSTables?

Reply:
I guess for a full rebuild, we would have already thrown an error. `continue` should be harmless.

Reply:
I will include this on my list of testing deficiencies to look at because it would be nice to know the answer. More to the point, I'll check whether any of our existing tests cover this scenario.